pkg/server/sandbox_list.go

```diff
@@ -17,14 +17,129 @@ limitations under the License.
 package server
 
 import (
-	"errors"
+	"fmt"
 
+	"github.com/golang/glog"
 	"golang.org/x/net/context"
 
+	"github.com/containerd/containerd/api/services/execution"
+
+	"github.com/containerd/containerd/api/types/container"
 	"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+
+	"github.com/kubernetes-incubator/cri-containerd/pkg/metadata"
 )
 
 // ListPodSandbox returns a list of Sandbox.
-func (c *criContainerdService) ListPodSandbox(ctx context.Context, r *runtime.ListPodSandboxRequest) (*runtime.ListPodSandboxResponse, error) {
-	return nil, errors.New("not implemented")
+func (c *criContainerdService) ListPodSandbox(ctx context.Context, r *runtime.ListPodSandboxRequest) (retRes *runtime.ListPodSandboxResponse, retErr error) {
+	glog.V(4).Infof("ListPodSandbox with filter %+v", r.GetFilter())
+	defer func() {
+		if retErr == nil {
+			glog.V(4).Infof("ListPodSandbox returns sandboxes %+v", retRes.GetItems())
+		}
+	}()
+
+	// List all sandbox metadata from store.
+	sandboxesInStore, err := c.sandboxStore.List()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list metadata from sandbox store: %v", err)
+	}
+
+	resp, err := c.containerService.List(ctx, &execution.ListRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list sandbox containers: %v", err)
+	}
+	sandboxesInContainerd := resp.Containers
+
+	var sandboxes []*runtime.PodSandbox
+	for _, sandboxInStore := range sandboxesInStore {
+		meta, err := sandboxInStore.GetMetadata()
+		if err != nil {
+			// This should not happen, return an error.
+			return nil, fmt.Errorf("failed to get sandbox metadata: %v", err)
+		}
+
+		var sandboxInContainerd *container.Container
+		for _, s := range sandboxesInContainerd {
+			if s.ID == meta.ID {
+				sandboxInContainerd = s
+				break
+			}
+		}
+
+		// Set sandbox state to NOTREADY by default.
+		state := runtime.PodSandboxState_SANDBOX_NOTREADY
+		// If the sandbox container is running, return the sandbox as READY.
+		if sandboxInContainerd != nil && sandboxInContainerd.Status == container.Status_RUNNING {
+			state = runtime.PodSandboxState_SANDBOX_READY
+		}
+
+		sandboxes = append(sandboxes, toCRISandbox(meta, state))
+	}
+
+	sandboxes = c.filterCRISandboxes(sandboxes, r.GetFilter())
+	return &runtime.ListPodSandboxResponse{Items: sandboxes}, nil
+}
+
+// toCRISandbox converts sandbox metadata into CRI pod sandbox.
+func toCRISandbox(meta metadata.SandboxMetadata, state runtime.PodSandboxState) *runtime.PodSandbox {
+	return &runtime.PodSandbox{
+		Id:          meta.ID,
+		Metadata:    meta.Config.GetMetadata(),
+		State:       state,
```

Comment:
> I'm surprised that state is only ready or not ready and that the actual container state is tossed... this will force users to use runc or other debug tool to find out actual state. We should discuss and maybe open an issue if one does not already exist.

In CRI, we only care about whether the sandbox is ready/notready. If it's ready, we'll keep using it; if not, we'll recreate it. We don't care about any other status for now.

Currently, the sandbox is ready when pause container is running, not ready if not. This is also how `dockershim` is behaving.